### PR TITLE
CloudWatch EMF Metric: Add support for cpu and mem input metrics

### PR DESCRIPTION
**Pull request by hossain-rayhan**
This changes will enable customers to send cpu and mem metrics to CloudWatch in EMF format.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: include/fluent-bit/flb_intermediate_metric.h
Comment (edsiper) [inline include/fluent-bit/flb_intermediate_metric.h:1]: - append Fluent Bit licensing comments: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#licensing

Comment (edsiper) [reply include/fluent-bit/flb_intermediate_metric.h:1]: every header file content must be inside a conditional macro, e.g:

```c
#ifndef FLB_INTERMEDIATE_METRIC
#define FLB_INTERMEDIATE_METRIC
...
#endif

note: I am not confident the name of this file is the proper one
```

File: include/fluent-bit/flb_intermediate_metric.h
Comment (edsiper) [inline include/fluent-bit/flb_intermediate_metric.h:6]: do no use double forward slash: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#single-line-comment

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:530]: do not exceed 80 characters in one line

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:557]: avoid nested calls, calc the size outside of the function

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:583]: add brace after else: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:594]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:597]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:611]: if I am not wrong csv_values is going out of the scope leading to a memory leak

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:609]: the flb_split_entry structure already has a ```len``` property, the strlen() is not needed.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:712]: do not use printf, use flb_plg_info() instead

**Comment by hossain-rayhan** [comment]
Hi @edsiper , thanks for your fast and detailed feedback- it was awesome. In the above [commit](https://github.com/fluent/fluent-bit/pull/2187/commits/66df57d335620a4e83416717274d7f2d8f5c0024), I tried to follow the fluent-bit guidelines and address your feedbacks. Will wait for another round of review from you.

File: include/fluent-bit/flb_intermediate_metric.h
Comment (PettitWesley) [inline include/fluent-bit/flb_intermediate_metric.h:37]: What does the "ir" stand for

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:705]: nit: since you have the literal string "cpu" here, not sure how useful it is to have a separate constant for the length of the string... probably more readable just to put the number 3 here

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (PettitWesley) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c:615]: I think you're right, @hossain-rayhan `flb_utils_split` allocates memory in each iteration of the loop...

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:713]: Why return here? If the input plugin is not for metrics.. then don't you want to continue and process the msgpack as logs like normal?

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:791]: where is this memory freed?

**Comment by PettitWesley** [comment]
Please test the plugin and run Valgrind, I think there are some memory leaks

**Comment by hossain-rayhan** [comment]
@PettitWesley and @edsiper , thanks for your feedback. I added another [commit] (https://github.com/fluent/fluent-bit/pull/2187/commits/af6871eeac8c5a34bdca60107c8ef756bc3ea531) to address all of them, and I tested my code with Valgrind. Feel free to throw some comments.

**Comment by PettitWesley** [comment]
@hossain-rayhan Please squah

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:528]: ```cpu_p``` string length is 5, not 4.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:529]: fix indentation

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:562]: variables must be defined on top

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:560]: packer name must be ```mp_pck``` and sbuffer ```mp_sbuf``` (we use the same name in core and plugins)

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:634]: declare variables on top of the function

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:707]: ```else if``` must be in the next line

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (edsiper) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:787]: after every memory allocation, check the return value, if is null raise an ```flb_errno()``` and return.

**Comment by PettitWesley** [comment]
@hossain-rayhan Please diff this change with master. The CloudWatch Logs plugin has been merged.

File: include/fluent-bit/flb_intermediate_metric.h
Comment (hossain-rayhan) [reply include/fluent-bit/flb_intermediate_metric.h]: Updated.

File: include/fluent-bit/flb_intermediate_metric.h
Comment (hossain-rayhan) [reply include/fluent-bit/flb_intermediate_metric.h]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Addressed.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c:710]: Memory freed. I think it should help.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: include/fluent-bit/flb_intermediate_metric.h
Comment (hossain-rayhan) [reply include/fluent-bit/flb_intermediate_metric.h]: Renamed.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: I think `strncmp` compares the first n chars regardless the length of the string. I updated it.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c:880]: Updated.

**Comment by PettitWesley** [comment]
@hossain-rayhan Please adjust your commit messages to fit the style guide: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#commit-changes

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:744]: nit: spelling mistake

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:884]: for all error cases, before returning, you need to clean up and free memory

in this function, the error label takes care of that; so use `goto error` like the other code does

File: plugins/out_cloudwatch_logs/cloudwatch_logs.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_logs.c:412]: I think this code was copied by mistake- `buf` is allocated in `cb_cloudwatch_init` and set on the context. It should not be recreated in the flush method. 

See in the existing code, `ctx->buf` is passed: https://github.com/fluent/fluent-bit/blob/master/plugins/out_cloudwatch_logs/cloudwatch_logs.c#L370

File: plugins/out_cloudwatch_logs/cloudwatch_logs.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_logs.c:582]: You comment inside of the `flb_cloudwatch` structure was better. Please copy that or write a longer similar explanation here that will allow the user to understand the lists of lists idea.

**Comment by hossain-rayhan** [comment]
Valgrind log output attached. 

Screen Shot 2020-10-02 at 10 24 31 AM

File: plugins/out_cloudwatch_logs/cloudwatch_logs.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_logs.c]: Yea, I messed it up while rebasing with `master`. Copied the original code from `master`.

File: plugins/out_cloudwatch_logs/cloudwatch_logs.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_logs.c]: Updated.

File: plugins/out_cloudwatch_logs/cloudwatch_api.c
Comment (hossain-rayhan) [reply plugins/out_cloudwatch_logs/cloudwatch_api.c]: Updated.

**Comment by hossain-rayhan** [comment]
@PettitWesley  and @edsiper, can I have another review. I added the `valgrind` log output.

**Comment by PettitWesley** [comment]
@hossain-rayhan For Valgrind, we want to see the output once the program is stopped and exits, that's when it will print info about leaks. I don't see that in your screenshot

File: plugins/out_cloudwatch_logs/cloudwatch_logs.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_logs.c:553]: Spelling mistake: **"list of lists"**.

File: src/CMakeLists.txt
Comment (PettitWesley) [inline src/CMakeLists.txt:138]: Something went wrong here- there is already an if block a few lines above for `FLB_AWS`. Remove this one, it is not needed.

**Comment by PettitWesley** [comment]
@hossain-rayhan please submit a PR to update the docs here: https://github.com/fluent/fluent-bit-docs

I think this is almost ready to merge.

File: src/CMakeLists.txt
Comment (hossain-rayhan) [reply src/CMakeLists.txt]: removed.

**Review by PettitWesley — APPROVED** [review]

File: plugins/out_cloudwatch_logs/cloudwatch_logs.c
Comment (PettitWesley) [inline plugins/out_cloudwatch_logs/cloudwatch_logs.c:552]: Spelling: "list of lists"

**Comment by hossain-rayhan** [comment]
Documentation PR: https://github.com/fluent/fluent-bit-docs/pull/395

**Comment by hossain-rayhan** [comment]
For the Valgrind, I cannot see any log traces for definitely lost blocks. Also, for the possibly lost, I don't think they are coming from my piece of code. Here, I am attaching a screenshot from valgrind log trace for better visibility.


Screen Shot 2020-10-07 at 12 23 59 PM

**Review by PettitWesley — APPROVED** [review]

**Comment by edsiper** [comment]
thanks.

Merging this PR. Once we support fully metrics support we should deprecate the intermediate interface added here.

**Pull request merged**
